rsmtool/utils.py

```diff
@@ -506,62 +506,228 @@ def agreement(score1, score2, tolerance=0):
 
     Returns
     -------
     smd : float
         The SMD for the given group or subgroup.
 
+    Raises
+    ------
+    ValueError
+        If method='williamson' and either population_y_true_sd or
+        population_y_pred_sd is None.
+    ValueError
+        If method is not in {'unpooled', 'pooled', 'williamson', 'johnson'}.
+
     Notes
     -----
-    The current implementation was recommended by Williamson, et al. (2012).
+    The 'williamson' implementation was recommended by Williamson, et al. (2012).
     The metric is only applicable when both sets of scores are on the same
-    scale. Additional implementations may be added in the future.
+    scale.
     """
+    y_true_avg = np.mean(y_true)
+    y_pred_avg = np.mean(y_pred)
+    numerator = y_true_avg - y_pred_avg
+
     method = method.lower()
-    if method == 'williamson':
-        numerator = mean_system_score - mean_human_score
-        denominator = np.sqrt((population_system_score_sd**2 +
-                               population_human_score_sd**2) / 2)
+    if method == 'unpooled':
+        denominator = np.std(y_true, ddof=ddof)
+    elif method == 'pooled':
+        denominator = np.sqrt((np.std(y_true, ddof=ddof)**2 +
+                               np.std(y_pred, ddof=ddof)**2) / 2)
+    elif method == 'johnson':
+        if population_y_true_sd is None:
+            raise ValueError("If `method='johnson'`, then `population_y_true_sd` "
+                             "must be provided.")
+        denominator = population_y_true_sd
+    elif method == 'williamson':
+        if population_y_true_sd is None or population_y_pred_sd is None:
+            raise ValueError("If `method='williamson'`, both `population_y_true_sd` "
+                             "and `population_y_pred_sd` must be provided.")
+        numerator = y_pred_avg - y_true_avg
+        denominator = np.sqrt((population_y_true_sd**2 +
+                               population_y_pred_sd**2) / 2)
     else:
-        raise ValueError("Currently, only the 'williamson' SMD method is "
-                         "supported. You specified: {}".format(method))
+        possible_methods = {"'unpooled'", "'pooled'", "'johnson'", "'williamson'"}
+        raise ValueError("The available methods are {{{}}}; you selected {}."
+                         "".format(', '.join(possible_methods), method))
 
     # if the denominator is zero, then return NaN as the SMD
     smd = np.nan if denominator == 0 else numerator / denominator
     return smd
 
 
+def difference_of_standardized_means(y_true,
+                                     y_pred,
+                                     population_y_true_mn=None,
+                                     population_y_pred_mn=None,
+                                     population_y_true_sd=None,
+                                     population_y_pred_sd=None,
+                                     ddof=1):
+    """
+    Calculate the difference of standardized means.
```

Comment:
"This is done by standardizing both observed and predicted scores to z-scores using mean and std for the whole population and then calculating differences between standardized means for each subgroup"